orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -825,6 +819,206 @@ void AddNewScalarNodeArgAndInitializer(Graph& graph,
+    current_node_arg = &new_node_arg;
+
+    // process recv node in cut i
+    auto& recv_attributes = recv_nodes[i]->GetMutableAttributes();
+    auto& recv_element_types = recv_attributes["element_types"];
+    recv_element_types.add_ints(static_cast<int64_t>(dtype));
+    recv_nodes[i]->MutableOutputDefs().push_back(current_node_arg);
+  }
+
+  // update the consumer node's input if the node's group is not in the first partition
+  for (size_t i = 1u; i < node_groups.size(); ++i) {
+    ORT_ENFORCE(node_groups[i].stage_id > from_stage, "node group id (", node_groups[i].stage_id,
+                ") is less than first stage id (", from_stage, "). ");
+    size_t new_node_arg_index = node_groups[i].stage_id - from_stage - 1;
+    for(auto node : node_groups[i].nodes){
+      for (auto& input_node : node->MutableInputDefs()) {
+        if (input_node == initializer) {
+          input_node = new_node_args[new_node_arg_index];
+          break;
+        }
+      }
+    }
+  }
+  return Status::OK();
+}
+
+void TraverseGraphWithConnectedElement(Graph& graph,
+                                       const Node* start_node,
+                                       std::set<const Node*>& visited_nodes,
+                                       std::set<const NodeArg*>& visited_inputs,
+                                       std::set<const NodeArg*>& visited_outputs) {
+  visited_nodes.clear();
+  visited_inputs.clear();
+  visited_outputs.clear();
+
+  std::queue<const Node*> node_queue;
+  node_queue.push(start_node);
+
+  while (!node_queue.empty()) {
+    auto node = node_queue.front();
+    node_queue.pop();
+    if (visited_nodes.insert(node).second) {
+      std::vector<const Node*> connected_nodes;
+      ORT_THROW_IF_ERROR(FindAllConnectedNodes(graph, node, connected_nodes, visited_inputs, visited_outputs));
+
+      for (auto n : connected_nodes) {
+        ORT_ENFORCE(n != nullptr, "Found nullptr in searching for connected nodes");
+        node_queue.push(n);
+      }
+    }
+  }
+}
+
+// If an initializer is shared across partitions, instead of creating a separate all_reduce op to
+// sync with those tensors in selected partitions, we save only one copy of that initializer in
+// the very first partition it appears, and pass that data down to all following partitions
+// where this initializer is used.
+common::Status HandleSharedInitializer(Graph& graph,
+                                       const std::vector<Node*>& send_nodes,
+                                       const std::vector<Node*>& recv_nodes) {
+  // Map a given initializer to all the partitions that its consumer nodes reside. The size of
+  // the mapped vector reflects how many partitions this initializer's consumer nodes distribute.
+  // If its size is greater than 1, it means this initializer is being used in more than one partition and
+  // we need to proceed those cases.
+  std::map<const NodeArg*, std::vector<PipelineStageNodeGroup>> input_consumer_stage_map;
```

Comment:
>input_consumer_stage_map [](start = 64, length = 24)

So what's the current status of the argument `Graph& graph`? Is it the original graph? Or the original graph + Send/Recv's before partition? Looks like we need sub-graphs from all stages to construct this map.

Reply:
Yep. At this point, no nodes have been removed yet. It is the modified original graph + send/recv.

---
In reply to: [452402499](https://github.com/microsoft/onnxruntime/pull/4321#discussion_r452402499) [](ancestors = 452402499)